File: Dynamic_Programming/GridTravelerMinPath.java
/*
Given a m x n grid filled with non-negative numbers, find a path from top left
to bottom right, which minimizes the sum of all numbers along its path.

Note: You can only move either down or right at any point in time.

Input: grid = [[1,3,1],[1,5,1],[4,2,1]]
Output: 7
Explanation: Because the path 1 → 3 → 1 → 1 → 1 minimizes the sum.
 */

 public class GridTravelerMinPath {

     public int minPath(int[][] grid) {
         int m = grid.length;
         int n = grid[0].length;

         for (int i = 1; i < n; i += 1) {
             grid[0][i] += grid[0][i - 1];
         }
         for (int i = 1; i < m; i += 1) {
             grid[i][0] += grid[i - 1][0];
         }

         for (int i = 1; i < m; i += 1) {
             for (int j = 1; j < n; j += 1) {
                 int min = Math.min(grid[i - 1][j], grid[i][j - 1]);
                 grid[i][j] += min;
             }
         }
         return grid[m - 1][n - 1];
     }

     public static void main(String[] agrs) {
         GridTravelerMinPath gtmp = new GridTravelerMinPath();
         int[][] path1 = {
             {1,1,1},
             {1,5,5},
             {1,6,5}
         };
         int[][] path2 = {{1,3,1}};
         int[][] path3 = {{1}};

         System.out.println(gtmp.minPath(path1));
         System.out.println(gtmp.minPath(path2));
         System.out.println(gtmp.minPath(path3));
     }
 }

File: Dynamic_Programming/GridTravelerMinPath.py
"""
Given a m x n grid filled with non-negative numbers, find a path from top left
to bottom right, which minimizes the sum of all numbers along its path.

Note: You can only move either down or right at any point in time.

Input: grid = [[1,3,1],[1,5,1],[4,2,1]]
Output: 7
Explanation: Because the path 1 -> 3 -> 1 -> 1 -> 1 minimizes the sum.
"""

def minPath(grid):
    m = len(grid)
    n = len(grid[0])

    for i in range(m):
        for j in range(n):
            if (i == 0 and j == 0):
                pass
            elif (i == 0):
                grid[i][j] += grid[i][j - 1]
            elif (j == 0):
                grid[i][j] += grid[i - 1][j]
            else:
                minPath = min(grid[i - 1][j], grid[i][j - 1])
                grid[i][j] += minPath
    return grid[m - 1][n - 1]


print(minPath([[1,1,1], [1,5,5], [1,6,5]]))
print(minPath([[1,3,1]]))
print(minPath([[1]]))
